io/shared/src/main/scala/org/scalajs/io/MemFiles.scala

```diff
@@ -10,52 +10,49 @@
 package org.scalajs.io
 
 import java.io._
-
 import java.nio.charset.StandardCharsets
 
-/** A base class for simple in-memory mutable virtual files. */
-class MemVirtualFile(val path: String) extends VirtualFile {
-  private[this] var _version: Option[String] = None
-
-  override def version: Option[String] = _version
-  def version_=(v: Option[String]): Unit = _version = v
-
-  final def withVersion(v: Option[String]): this.type = {
-    version = v
-    this
-  }
+/** A simple in-memory immutable virtual binary file. */
+class MemVirtualBinaryFile(
+    final val path: String,
+    content: Array[Byte],
+    final override val version: Option[String]
+) extends VirtualBinaryFile {
+  final def inputStream: InputStream = new ByteArrayInputStream(content)
 }
 
-/** A simple in-memory mutable virtual binary file. */
-class MemVirtualBinaryFile(p: String) extends MemVirtualFile(p)
-                                         with VirtualBinaryFile {
-  private[this] var _content: Array[Byte] = new Array[Byte](0)
+object MemVirtualBinaryFile {
+  def apply(path: String, content: Array[Byte]): MemVirtualBinaryFile =
+    apply(path, content, None)
 
-  def content: Array[Byte] = _content
-  def content_=(v: Array[Byte]): Unit = _content = v
+  def apply(path: String, content: Array[Byte],
+      version: Option[String]): MemVirtualBinaryFile = {
+    new MemVirtualBinaryFile(path, content, version)
```

Comment:
Hum ... if we're being pedantic, we would still need a defensive copy of `content` when allocating the `MemVirtualBinaryFile`. Otherwise the caller could still modify the `content` afterwards if it keeps a reference to it :s

Reply:
I have declared this as UB :) Making a copy is probably not a good idea because this is of the path of reading jars.